specs/asgi.rst

```diff
@@ -143,46 +143,29 @@ serializable, and so they are only allowed to contain the following types:
 Applications
 ------------
 
-ASGI applications are defined as a callable::
+.. note::
 
-    application(scope)
-
-* ``scope``: The Connection Scope, a dictionary that contains at least a
-  ``type`` key specifying the protocol that is incoming.
+    The application format changed in 3.0 to use a single callable, rather than
+    the prior two-callable format. Two-callable is documented below in
+    "Legacy Applications"; servers can easily implement support for it using
+    the ``asgiref.compatibility`` library, and should try to support it.
```

Comment:
I'm not sure we should try to support both, I'm thinking it would be a pain specifically in middleware to do so. I'm tempted to take advantage of some breaking changes in Hypercorn to switch to asgi 3 only.

Reply:
>  I'm thinking it would be a pain specifically in middleware to do so.

No doubt. I think the note's more intended at server implementations. I don't think it's a particularly big deal if Hypercorn prefers to go 3.0 only, so long as there's alternatives providing a graceful 2-or-3 policy. (I'll plan to go 2-or-3 with Uvicorn)

Reply:
Actually, I'm happy with this wording. I think I'll support both in Hypercorn.